src/useObservable.ts

```diff
@@ -33,24 +38,40 @@ export function useObservable<T>(observable: Observable<T>): T | undefined
 export function useObservable<T>(observable: Observable<T>, initialValue: T): T
 export function useObservable<T>(observable: Observable<T>, initialValue: () => T): T
 export function useObservable<T>(observable: Observable<T>, initialValue?: T | (() => T)) {
-  const [getSnapshot, subscribe] = useMemo(() => {
+  const [getSnapshot, subscribe] = useMemo<
+    [() => T, Parameters<typeof useSyncExternalStore>[0]]
+  >(() => {
     const record = getOrCreateStore(observable, getValue(initialValue))!
     return [
       function getSnapshot() {
+        // @TODO: perf opt opportunity: we could do `record.subscription.unsubscribe()` here to clear up some memory, as this subscription is only needed to provide a sync initialValue.
         return record.currentValue
       },
-      function subscribe(callback: (value: T) => void) {
-        const sub = record.observable.subscribe(next => callback(next))
+      function subscribe(callback: () => void) {
+        // @TODO: perf opt opportunity: we could do `record.subscription.unsubscribe()` here as we only need 1 subscription active to keep the observer alive
+        const sub = record.observable.subscribe(() => callback())
         return () => {
           sub.unsubscribe()
         }
       },
     ]
   }, [observable])
 
-  useIsomorphicEffect(() => {
+  const shouldRestoreSubscriptionRef = useRef(false)
+  useEffect(() => {
+    const store = getOrCreateStore(observable, getValue(initialValue))!
+    if (shouldRestoreSubscriptionRef.current) {
+      if (store.subscription.closed) {
+        store.subscription = store.observable.subscribe()
+      }
+      shouldRestoreSubscriptionRef.current = false
+    }
+
     return () => {
-      getOrCreateStore(observable, getValue(initialValue))!.subscription.unsubscribe()
+      // React StrictMode will call effects as `setup + teardown + setup` thus we can't trust this callback as "react is about to unmount"
+      // Tracking this ref lets us set the subscription back up on the next `setup` call if needed, and if it really did unmounted then all is well
+      shouldRestoreSubscriptionRef.current = !store.subscription.closed
+      store.subscription.unsubscribe()
```

Comment:
Using `store` from the setup function, instead of doing a `getOrCreateStore(observable, getValue(initialValue))` like before in teardown, it's ensured that if `observable` changes identity we don't "forget" to unsubscribe old stores and leak memory.